### DISCUSSION TITLE
Anvil not working

### DISCUSSION THREAD

**Discussion by Abhishek-2416** [Cyfrin/foundry-full-course-cu#1434] [Q&A]
I was doing the FundMe contract and i am able to run all my test on the SEPOLIA network but not on the anvil or local chain
Unable to figure out what is going wrong Please help me out here
https://github.com/Abhishek-2416/Foundry-FundMe

**Comment by note-xin** [comment]
hey, there friend.
Could you show the error report?

**Reply by Abhishek-2416** [reply]
Running 1 test for test/uint/FundMeTest.t.sol:FundMeTest
[FAIL. Reason: EvmError: Revert] testOnlyOwnerCanDoCheaperWithdraw() (gas: 22157)
Traces:
  [3034855] FundMeTest::setUp()
    ├─ [1514679] → new DeployFundMe@0x5615dEB798BB3E4dFa0139dFa1b3D433Cc23b72f
    │   └─ ← 7455 bytes of code
    ├─ [1461296] DeployFundMe::run()
    │   ├─ [903342] → new HelperConfig@0x104fBc016F4bb334D775a19E8A6510109AC63E00
    │   │   ├─ [0] VM::startBroadcast()
    │   │   │   └─ ← ()
    │   │   ├─ [377261] → new MockV3Aggregator@0x90193C961A926261B756D1E5bb255e67ff9498A1
    │   │   │   └─ ← 1107 bytes of code
    │   │   ├─ [0] VM::stopBroadcast()
    │   │   │   └─ ← ()
    │   │   ├─ emit HelperConfig__CreatedMockPriceFeed(priceFeed: MockV3Aggregator: [0x90193C961A926261B756D1E5bb255e67ff9498A1])
    │   │   └─ ← 2222 bytes of code
    │   ├─ [381] HelperConfig::activeNetworkConfig() [staticcall]
    │   │   └─ ← MockV3Aggregator: [0x90193C961A926261B756D1E5bb255e67ff9498A1]
    │   ├─ [0] VM::startBroadcast()
    │   │   └─ ← ()
    │   ├─ [490940] → new FundMe@0xA8452Ec99ce0C64f20701dB7dD3abDb607c00496
    │   │   └─ ← 2340 bytes of code
    │   ├─ [0] VM::stopBroadcast()
    │   │   └─ ← ()
    │   └─ ← FundMe: [0xA8452Ec99ce0C64f20701dB7dD3abDb607c00496]
    ├─ [0] VM::deal(0x0000000000000000000000000000000000000001, 10000000000000000000 [1e19])
    │   └─ ← ()
    └─ ← ()

  [22157] FundMeTest::testOnlyOwnerCanDoCheaperWithdraw()
    ├─ [0] VM::prank(0x0000000000000000000000000000000000000001)
    │   └─ ← ()
    ├─ [5200] FundMe::fund{value: 1000000000000000000}()
    │   ├─ [0] 0x694AA1769357215DE4FAC081bf1f309aDC325306::latestRoundData() [staticcall]
    │   │   └─ ← ()
    │   └─ ← EvmError: Revert
    └─ ← EvmError: Revert

Test result: FAILED. 0 passed; 1 failed; 0 skipped; finished in 769.01µs
 
Ran 1 test suites: 0 tests passed, 1 failed, 0 skipped (1 total tests)

Failing tests:
Encountered 1 failing test in test/uint/FundMeTest.t.sol:FundMeTest
[FAIL. Reason: EvmError: Revert] testOnlyOwnerCanDoCheaperWithdraw() (gas: 22157)

Encountered a total of 1 failing tests, 0 tests succeeded

**Comment by note-xin** [comment]
your test function
```solidity
function testOnlyOwnerCanDoCheaperWithdraw() public funded {
        //Arrange
        uint160 numberOfFunders = 10;
        uint160 startingFunderIndex = 1;
        for (uint160 i = startingFunderIndex; i < numberOfFunders; i++) {
            hoax(address(i), STARTING_BALANCE);
            fundMe.fund{value: FUND_VALUE}();
        }

        uint256 startingOwnerBalance = fundMe.getOwner().balance;

        //Act
        vm.startPrank(fundMe.getOwner());
        fundMe.cheaperWithdraw();
        vm.stopPrank();

        //Assert
        uint256 endingOwnerBalance = fundMe.getOwner().balance;
        uint256 endingFundMeBalance = address(fundMe).balance;

        assertEq(endingFundMeBalance, 0);
        assertEq(endingOwnerBalance, startingOwnerBalance + (FUND_VALUE * numberOfFunders));
    }
```

the error occurs when you try to call your `fund` function in the `funded` modifier.
try lowering your `FUND_VALUE` to `0.2 ether` or double-check your `anvilEthConfig` and revert condition of your `fund` function

try `-vvvvv` for more info on error

**Reply by Abhishek-2416** [reply]
This is my error report when i reduce the amount which is to be funded, and yes i have checked my anvilEthConfig from patricks code
But all my test cases are passing when i am using SEPOLIA and these arent when on anvil
So its either problem with the HelperConfig or either the Mock contract

Encountered 1 failing test in test/uint/FundMeTest.t.sol:FundMeTest
[FAIL. Reason: EvmError: Revert] testFundUpdatesFundedDataStructure() (gas: 22113)

Encountered a total of 1 failing tests, 0 tests succeeded
[main][~/Desktop/Projects/Patrick Collins/Foundry FundMe]$ forge test --mt testFundUpdatesFundedDataStructure -vvvvv
[⠢] Compiling...
[⠑] Compiling 1 files with 0.8.21
[⠘] Solc 0.8.21 finished in 1.42s
Compiler run successful!

Running 1 test for test/uint/FundMeTest.t.sol:FundMeTest
[FAIL. Reason: EvmError: Revert] testFundUpdatesFundedDataStructure() (gas: 22113)
Traces:
  [3034855] FundMeTest::setUp()
    ├─ [1514679] → new DeployFundMe@0x5615dEB798BB3E4dFa0139dFa1b3D433Cc23b72f
    │   └─ ← 7455 bytes of code
    ├─ [1461296] DeployFundMe::run()
    │   ├─ [903342] → new HelperConfig@0x104fBc016F4bb334D775a19E8A6510109AC63E00
    │   │   ├─ [0] VM::startBroadcast()
    │   │   │   └─ ← ()
    │   │   ├─ [377261] → new MockV3Aggregator@0x90193C961A926261B756D1E5bb255e67ff9498A1
    │   │   │   └─ ← 1107 bytes of code
    │   │   ├─ [0] VM::stopBroadcast()
    │   │   │   └─ ← ()
    │   │   ├─ emit HelperConfig__CreatedMockPriceFeed(priceFeed: MockV3Aggregator: [0x90193C961A926261B756D1E5bb255e67ff9498A1])
    │   │   └─ ← 2222 bytes of code
    │   ├─ [381] HelperConfig::activeNetworkConfig() [staticcall]
    │   │   └─ ← MockV3Aggregator: [0x90193C961A926261B756D1E5bb255e67ff9498A1]
    │   ├─ [0] VM::startBroadcast()
    │   │   └─ ← ()
    │   ├─ [490940] → new FundMe@0xA8452Ec99ce0C64f20701dB7dD3abDb607c00496
    │   │   └─ ← 2340 bytes of code
    │   ├─ [0] VM::stopBroadcast()
    │   │   └─ ← ()
    │   └─ ← FundMe: [0xA8452Ec99ce0C64f20701dB7dD3abDb607c00496]
    ├─ [0] VM::deal(0x0000000000000000000000000000000000000001, 10000000000000000000 [1e19])
    │   └─ ← ()
    └─ ← ()

  [22113] FundMeTest::testFundUpdatesFundedDataStructure()
    ├─ [0] VM::prank(0x0000000000000000000000000000000000000001)
    │   └─ ← ()
    ├─ [5200] FundMe::fund{value: 200000000000000000}()
    │   ├─ [0] 0x694AA1769357215DE4FAC081bf1f309aDC325306::latestRoundData() [staticcall]
    │   │   └─ ← ()
    │   └─ ← EvmError: Revert
    └─ ← EvmError: Revert

Test result: FAILED. 0 passed; 1 failed; 0 skipped; finished in 859.24µs
 
Ran 1 test suites: 0 tests passed, 1 failed, 0 skipped (1 total tests)

Failing tests:
Encountered 1 failing test in test/uint/FundMeTest.t.sol:FundMeTest
[FAIL. Reason: EvmError: Revert] testFundUpdatesFundedDataStructure() (gas: 22113)

Encountered a total of 1 failing tests, 0 tests succeeded

**Comment by EngrPips** [comment]
@Abhishek-2416  I hope the help provided by @note-xin  helped you resolved your issue.

**Comment by note-xin** [comment]
go to your `deployment script`
```solidity
contract DeployFundMe is Script {
    //We need to start all the deploy contracts with the "run" function
    function run() external returns (FundMe) {
        HelperConfig helperConfig = new HelperConfig();
        address ethUsdPriceFeed = helperConfig.activeNetworkConfig();

        vm.startBroadcast();
        // We will now create a MOCK contract
        FundMe fundMe = new FundMe(ethUsdPriceFeed);
        vm.stopBroadcast();
        return fundMe;
    }
}
```
make it return your network script as well. like

```solidity
contract DeployFundMe is Script {
    //We need to start all the deploy contracts with the "run" function
    function run() external returns (FundMe,  HelperConfig) {
        HelperConfig helperConfig = new HelperConfig();
        address ethUsdPriceFeed = helperConfig.activeNetworkConfig();

        vm.startBroadcast();
        // We will now create a MOCK contract
        FundMe fundMe = new FundMe(ethUsdPriceFeed);
        vm.stopBroadcast();
        return (fundMe, HelperConfig);
    }
}
```
also, it will return a `tuple` so make the changes where ever you call `run`.
for example

```solidity
function setUp() external {
        //This is how the flow of the contract while testing works

        //We -> FundMeTest -> FundMe , Here we deploy the FundMeTest and this FundMeTest then deploys the FundMe Contract
        DeployFundMe deployFundMe = new DeployFundMe();
        (fundMe, helperconfig) = deployFundMe.run();

        vm.deal(bob, STARTING_BALANCE);
    }
``` 

but not sure if this will solve your issue ...

**Reply by Abhishek-2416** [reply]
Yeah this was a problem , but that didn't solve the anvil issue
